internal/builders/golang/targets.go

```diff
@@ -1,61 +1,77 @@
-package buildtarget
+package golang
 
 import (
+	"fmt"
+
 	"github.com/apex/log"
 	"github.com/goreleaser/goreleaser/config"
 )
 
-// All returns all valid build targets for a given build
-func All(build config.Build) (targets []Target) {
+type target struct {
+	os, arch, arm string
+}
+
+func (t target) String() string {
+	if t.arm != "" {
+		return fmt.Sprintf("%s_%s_%s", t.os, t.arch, t.arm)
+	}
+	return fmt.Sprintf("%s_%s", t.os, t.arch)
+}
+
+func matrix(build config.Build) (result []string) {
+	var targets []target
 	for _, target := range allBuildTargets(build) {
 		if !valid(target) {
-			log.WithField("target", target.PrettyString()).
+			log.WithField("target", target).
 				Debug("skipped invalid build")
 			continue
 		}
 		if ignored(build, target) {
-			log.WithField("target", target.PrettyString()).
+			log.WithField("target", target).
 				Debug("skipped ignored build")
 			continue
 		}
 		targets = append(targets, target)
 	}
+	for _, target := range targets {
+		result = append(result, target.String())
+	}
 	return
 }
 
-func allBuildTargets(build config.Build) (targets []Target) {
+func allBuildTargets(build config.Build) (targets []target) {
 	for _, goos := range build.Goos {
 		for _, goarch := range build.Goarch {
 			if goarch == "arm" {
 				for _, goarm := range build.Goarm {
-					targets = append(targets, New(goos, goarch, goarm))
+					targets = append(targets, target{goos, goarch, goarm})
 				}
 				continue
 			}
-			targets = append(targets, New(goos, goarch, ""))
+			targets = append(targets, target{goos, goarch, ""})
 		}
 	}
 	return
 }
 
-func ignored(build config.Build, target Target) bool {
+func ignored(build config.Build, target target) bool {
 	for _, ig := range build.Ignore {
```

Comment:
Instead of the iteration, we could also use a set for simpler code and faster lookups:

```go
package main

import "fmt"

type target struct {
	os, arch, arm string
}

func main() {

	ignoredTargets := map[target]bool{target{"foo", "bar", "baz"}: true, target{"bli", "bla", "blub"}: true}

	test_target := target{"foo", "bar", "baz"}
	_, ok := ignoredTargets[test_target] // check for existence
	if ok {
		fmt.Printf("%v ignored", test_target)
	}
}
```

Reply:
this could indeed be improved, I'll add a TODO for it